wgpu-core/src/indirect_validation/draw.rs

```diff
@@ -0,0 +1,984 @@
+
+    let info = crate::device::create_validator(
+        wgt::Features::PUSH_CONSTANTS,
+        wgt::DownlevelFlags::empty(),
+        naga::valid::ValidationFlags::all(),
+    )
+    .validate(&module)
+    .map_err(|inner| {
+        CreateShaderModuleError::Validation(naga::error::ShaderError {
+            source: src.to_string(),
+            label: None,
+            inner: Box::new(inner),
+        })
+    })?;
+    let hal_shader = hal::ShaderInput::Naga(hal::NagaShader {
+        module: alloc::borrow::Cow::Owned(module),
+        info,
+        debug_source: None,
+    });
+    let hal_desc = hal::ShaderModuleDescriptor {
+        label: None,
+        runtime_checks: wgt::ShaderRuntimeChecks::unchecked(),
+    };
+    let module = unsafe { device.create_shader_module(&hal_desc, hal_shader) }.map_err(
+        |error| match error {
+            hal::ShaderError::Device(error) => {
+                CreateShaderModuleError::Device(DeviceError::from_hal(error))
+            }
+            hal::ShaderError::Compilation(ref msg) => {
+                log::error!("Shader error: {}", msg);
+                CreateShaderModuleError::Generation
+            }
+        },
+    )?;
+
+    Ok(module)
+}
+
+fn create_validation_pipeline(
+    device: &dyn hal::DynDevice,
+    module: &dyn hal::DynShaderModule,
+    pipeline_layout: &dyn hal::DynPipelineLayout,
+    supports_indirect_first_instance: bool,
+) -> Result<Box<dyn hal::DynComputePipeline>, CreateIndirectValidationPipelineError> {
+    let pipeline_desc = hal::ComputePipelineDescriptor {
+        label: None,
+        layout: pipeline_layout,
+        stage: hal::ProgrammableStage {
+            module,
+            entry_point: "main",
+            constants: &hashbrown::HashMap::from([(
+                "supports_indirect_first_instance".to_string(),
+                f64::from(supports_indirect_first_instance),
+            )]),
+            zero_initialize_workgroup_memory: false,
+        },
+        cache: None,
+    };
+    let pipeline =
+        unsafe { device.create_compute_pipeline(&pipeline_desc) }.map_err(|err| match err {
+            hal::PipelineError::Device(error) => {
+                CreateComputePipelineError::Device(DeviceError::from_hal(error))
+            }
+            hal::PipelineError::Linkage(_stages, msg) => CreateComputePipelineError::Internal(msg),
+            hal::PipelineError::EntryPoint(_stage) => CreateComputePipelineError::Internal(
+                crate::device::ENTRYPOINT_FAILURE_ERROR.to_string(),
+            ),
+            hal::PipelineError::PipelineConstants(_, error) => {
+                CreateComputePipelineError::PipelineConstants(error)
+            }
+        })?;
+
+    Ok(pipeline)
+}
+
+fn create_bind_group_layout(
+    device: &dyn hal::DynDevice,
+    read_only: bool,
+    has_dynamic_offset: bool,
+    min_binding_size: wgt::BufferSize,
```

Comment:
Could we pass a `wgt::BindingType` instead of three separate arguments

Reply:
`wgt::BindingType` is an enum though, any of its variants could be passed in.